compiler_gym/service/compilation_session.py

```diff
@@ -0,0 +1,90 @@
+# Copyright (c) Facebook, Inc. and its affiliates.
+#
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+from pathlib import Path
+from typing import List, Optional, Tuple
+
+from compiler_gym.service.proto import (
+    Action,
+    ActionSpace,
+    Benchmark,
+    Observation,
+    ObservationSpace,
+)
+
+
+class CompilationSession:
+    """Base class for encapsulating an incremental compilation session.
+
+    To add support for a new compiler, subclass from this base and provide
+    implementations of the abstract methods, then call
+    :func:`create_and_run_compiler_service
+    <compiler_gym.service.runtime.create_and_run_compiler_service>` and pass in
+    your class type:
+
+    .. code-block:: python
+
+        from compiler_gym.service import CompilationSession
+        from compiler_gym.service import runtime
+
+        class MyCompilationSession(CompilationSession):
+            ...
+
+        if __name__ == "__main__":
+            runtime.create_and_run_compiler_service(MyCompilationSession)
+    """
+
+    compiler_version: str = ""
+    """The compiler version."""
+
+    action_spaces: List[ActionSpace] = []
+    """A list of action spaces describing the capabilities of the compiler."""
+
+    observation_spaces: List[ObservationSpace] = []
+    """A list of feature vectors that this compiler provides."""
+
+    def __init__(
+        self, working_dir: Path, action_space: ActionSpace, benchmark: Benchmark
+    ):
```

Comment:
Just curious: why doesn't this return a `status` like the c++ code?

Reply:
We [don't use exceptions in C++](https://google.github.io/styleguide/cppguide.html#Exceptions) (FWIW I don't have particularly strong opinions on the matter), the `CompilationSession` constructor in C++ is a safe default, and any of the three methods that can fail (init, compute feature, run action) return `Status`.

For Python, exceptions are fine and all of the runtime logic that interacts with `CompilationSession` will catch exceptions and set the relevant RPC error code